packages/react-dom/src/events/DOMModernPluginEventSystem.js

```diff
@@ -224,6 +227,45 @@ export function dispatchEventForPluginEventSystem(
     ) {
       return;
     }
+    let node = targetInst;
+
+    // The below logic attempts to work out if we need to change
+    // the target fiber to a different ancestor. We had similar logic
+    // in the legacy event system, except the big difference between
+    // systems is that the modern event system now has an event listener
+    // attached to each React Root and React Portal Root. Together,
+    // the DOM nodes representing these roots are the "rootContainer".
+    // To figure out which ancestor instance we should use, we traverse
+    // up the fiber tree from the target instance and attempt to find
+    // root boundaries that match that of our current "rootContainer".
+    // If we find that "rootContainer", we find the parent fiber
+    // sub-tree for that root and make that our ancestor instance.
+    while (true) {
+      if (node === null) {
+        return;
+      }
+      const tag = node.tag;
+      if (tag === HostRoot || tag === HostPortal) {
+        const container = node.stateNode.containerInfo;
+        if (container !== rootContainer && tag === HostPortal) {
+          return;
```

Comment:
I originally tried this but had to add this mess: https://github.com/facebook/react/pull/18215/files#diff-6b784480f6f3297c42bc2ee46e6f943R212

Because I think that this is not sufficient if the Portal happened to be a Portal from a different root then it shouldn't bail out since the portal won't be bubbling into that root. We should add a test for that.

The longer term fix is probably to have a unique `internalInstanceKey` for each root. That way we don't have to bother with collisions. It'd just be like each root is whole other library.